kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/storage/PartitionLog.java

```diff
@@ -0,0 +1,332 @@
+
+    @Data
+    @AllArgsConstructor
+    public static final class LogAppendInfo {
+        private Optional<Long> firstOffset;
+        private Long lastOffset;
+        private Integer shallowCount;
+        private Boolean offsetsMonotonic;
+        private Long lastOffsetOfFirstBatch;
+        private Integer validBytes;
+
+        public Long numMessages() {
+            return firstOffset.map(firstOffsetVal -> {
+                if (firstOffsetVal >= 0 && lastOffset >= 0) {
+                    return lastOffset - firstOffsetVal + 1;
+                }
+                return 0L;
+            }).orElse(0L);
+        }
+    }
+
+    public CompletableFuture<Long> appendRecords(final MemoryRecords records,
+                                                 final short version,
+                                                 final AppendRecordsContext appendRecordsContext) {
+        return append(records, version, false, appendRecordsContext);
+    }
+
+    /**
+     * Append this message to pulsar.
+     *
+     * This method will generally be responsible for assigning offsets to the messages,
+     * however if the assignOffsets=false flag is passed we will only check that the existing offsets are valid.
+     *
+     * @param records The log records to append
+     * @param version Inter-broker message protocol version
+     * @param ignoreRecordSize true to skip validation of record size
+     * @param appendRecordsContext See {@link AppendRecordsContext}
+     */
+    private CompletableFuture<Long> append(final MemoryRecords records,
+                                           final short version,
+                                           final boolean ignoreRecordSize,
+                                           final AppendRecordsContext appendRecordsContext) {
+        CompletableFuture<Long> appendFuture = new CompletableFuture<>();
+        RequestStats requestStats = appendRecordsContext.getRequestStats();
+        KafkaTopicManager topicManager = appendRecordsContext.getTopicManager();
+        final long beforeRecordsProcess = time.nanoseconds();
+        try {
+            final LogAppendInfo appendInfo =
+                    analyzeAndValidateRecords(records, version, topicPartition, ignoreRecordSize);
+
+            // trim any invalid bytes or partial messages before appending it to the on-disk log
+            MemoryRecords validRecords = trimInvalidBytes(records, appendInfo);
+            synchronized (lock) {
+                // Append Message into pulsar
+                final CompletableFuture<Optional<PersistentTopic>> topicFuture =
+                        topicManager.getTopic(fullPartitionName);
+                if (topicFuture.isCompletedExceptionally()) {
+                    topicFuture.exceptionally(e -> {
+                        appendFuture.completeExceptionally(e);
+                        return Optional.empty();
+                    });
+                    return appendFuture;
+                }
```

Comment:
When we refactor the code, please make sure the behavior is nearly the same as before. I see you copied Kafka's logic here, like calling `analyzeAndValidateRecords` and `trimInvalidBytes`, even the `lock`. Maybe some are right. However, we should not do that in this PR. This PR should migrate the logic from the current KoP source.

Reply:
Ok, I removed it.

Maybe I'll add this logic in ProducerStateManager's PR.